extensions/email/src/main/java/org/mule/extension/email/api/retriever/imap/IMAPOperations.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.extension.email.api.retriever.imap;
+
+import static javax.mail.Flags.Flag.DELETED;
+import static javax.mail.Flags.Flag.SEEN;
+import static javax.mail.Folder.READ_ONLY;
+import static javax.mail.Folder.READ_WRITE;
+import org.mule.extension.email.api.retriever.RetrieverConnection;
+import org.mule.extension.email.internal.EmailAttributes;
+import org.mule.extension.email.internal.operations.RetrieveOperation;
+import org.mule.extension.email.internal.operations.SetFlagOperation;
+import org.mule.runtime.api.message.MuleMessage;
+import org.mule.runtime.core.api.MuleContext;
+import org.mule.runtime.extension.api.annotation.param.Connection;
+import org.mule.runtime.extension.api.annotation.param.Optional;
+import org.mule.runtime.extension.api.annotation.param.UseConfig;
+
+import java.util.List;
+
+import javax.inject.Inject;
+
+/**
+ * Basic set of operations which perform on top the IMAP email protocol.
+ *
+ * @since 4.0
+ */
+public class IMAPOperations
+{
+
+    @Inject
+    private MuleContext context;
+
+    //TODO: ADD MATCHER.
+    /**
+     * List all the emails in the configured folder.
+     *
+     * @param connection  the corresponding {@link RetrieverConnection} instance.
+     * @return a {@link List} of {@link MuleMessage} carrying all the emails and it's corresponding attributes.
+     */
+    public List<MuleMessage<String, EmailAttributes>> retrieve(@UseConfig IMAPConfiguration config, @Connection RetrieverConnection connection)
+    {
+        return new RetrieveOperation().retrieve(connection.getOpenFolder(READ_ONLY), context, config.shouldReadContent());
```

Comment:
once you change the config-povider relationship, the command pattern you're implementing here is likely to become overkill. Any reason to disagree? In either case , all these XXOperation classes should be renamed to XXCommand, and those commands should be reusable instead of creating one each time.